src/v/cluster/rm_stm.cc

```diff
@@ -733,39 +733,62 @@ ss::future<tx_errc> rm_stm::do_abort_tx(
     co_return tx_errc::none;
 }
 
+raft::replicate_stages rm_stm::replicate_in_stages(
+  model::batch_identity bid,
+  model::record_batch_reader r,
+  raft::replicate_options opts) {
+    auto enqueued = ss::make_lw_shared<available_promise<>>();
+    auto f = enqueued->get_future();
+    auto replicate_finished
+      = do_replicate(bid, std::move(r), opts, enqueued)
+          .then([](result<raft::replicate_result> res) mutable { return res; });
+    return raft::replicate_stages(std::move(f), std::move(replicate_finished));
+}
+
 ss::future<result<raft::replicate_result>> rm_stm::replicate(
   model::batch_identity bid,
-  model::record_batch_reader b,
+  model::record_batch_reader r,
   raft::replicate_options opts) {
-    return _state_lock.hold_read_lock().then(
-      [this, bid, opts, b = std::move(b)](
-        ss::basic_rwlock<>::holder unit) mutable {
+    auto enqueued = ss::make_lw_shared<available_promise<>>();
+    return do_replicate(bid, std::move(r), opts, enqueued);
+}
+
+ss::future<std::error_code>
+rm_stm::transfer_leadership(std::optional<model::node_id> target) {
+    return _state_lock.hold_write_lock().then(
```

Comment:
nit: I think a side effect of doing this is delayed leadership transfer (and hence delayed balancing, especially with upcoming work in this area), essentially starvation from this lock if there are a lot of transactions are in progress, given the underlying semaphore is "fair". I don't know if that is going to be a big problem in practice when there is huge transactional workload, just thinking out loud.

Reply:
when it's fair write_lock gets it moment to shine when all read locks taken before the write lock is requested are expired, replication supports batching and lasts around 5ms so transfer_leadership is delayed just by several milliseconds so it isn't a problem